File: day8/part1and2/main.py
import math
from collections import defaultdict

def connect_pair(pair, junctions):
    a, b = pair['a'], pair['b']
    
    if a['circuit'] is None and b['circuit'] is None:
        new_circuit = object()
        a['circuit'] = b['circuit'] = new_circuit
    elif a['circuit'] is None or b['circuit'] is None:
        a['circuit'] = b['circuit'] = a['circuit'] or b['circuit']
    elif a['circuit'] != b['circuit']:
        old_circuit = b['circuit']
        new_circuit = a['circuit']
        for junction in junctions:
            if junction['circuit'] == old_circuit:
                junction['circuit'] = new_circuit
    
    return pair

def circuit_sizes(junctions):
    connected = [j for j in junctions if j['circuit'] is not None]
    
    circuits = defaultdict(list)
    for junction in connected:
        circuits[id(junction['circuit'])].append(junction)
    
    sizes = [len(circuit) for circuit in circuits.values()]
    return sorted(sizes, reverse=True)

def all_connected(junctions):
    circuit = junctions[0]['circuit']
    return circuit is not None and all(j['circuit'] == circuit for j in junctions)

def parse(input_text):
    lines = input_text.strip().split('\n')
    junctions = []
    
    for line in lines:
        x, y, z = map(int, line.split(','))
        junctions.append({'x': x, 'y': y, 'z': z, 'circuit': None})
    
    pairs = []
    for i in range(len(junctions)):
        for j in range(i + 1, len(junctions)):
            a = junctions[i]
            b = junctions[j]
            dist = math.sqrt(
                (a['x'] - b['x']) ** 2 +
                (a['y'] - b['y']) ** 2 +
                (a['z'] - b['z']) ** 2
            )
            pairs.append({'a': a, 'b': b, 'dist': dist})
    
    pairs.sort(key=lambda p: p['dist'])
    
    return junctions, pairs

def part1(input_text):
    times = 1000
    junctions, pairs = parse(input_text)
    
    for i in range(times):
        connect_pair(pairs[i], junctions)
    
    sizes = circuit_sizes(junctions)
    result = sizes[0] * sizes[1] * sizes[2]
    print(result)

def part2(input_text):
    junctions, pairs = parse(input_text)
    pair = None
    idx = 0
    
    while not all_connected(junctions):
        pair = connect_pair(pairs[idx], junctions)
        idx += 1
    
    ans = pair['a']['x'] * pair['b']['x']
    print(ans)

if __name__ == "__main__":
    with open("input.txt", "r") as f:
        input_data = f.read()
    
    part1(input_data)
    part2(input_data)

File: day8/part1and2/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <cmath>
#include <algorithm>
#include <map>
#include <memory>

struct Junction {
    int x, y, z;
    std::shared_ptr<int> circuit;
};

struct Pair {
    Junction* a;
    Junction* b;
    double dist;
};

void connectPair(Pair& pair, std::vector<Junction>& junctions) {
    Junction* a = pair.a;
    Junction* b = pair.b;
    
    if (!a->circuit && !b->circuit) {
        auto newCircuit = std::make_shared<int>(rand());
        a->circuit = b->circuit = newCircuit;
    } else if (!a->circuit || !b->circuit) {
        a->circuit = b->circuit = a->circuit ? a->circuit : b->circuit;
    } else if (a->circuit != b->circuit) {
        auto oldCircuit = b->circuit;
        auto newCircuit = a->circuit;
        for (auto& junction : junctions) {
            if (junction.circuit == oldCircuit) {
                junction.circuit = newCircuit;
            }
        }
    }
}

std::vector<int> circuitSizes(const std::vector<Junction>& junctions) {
    std::map<std::shared_ptr<int>, int> circuits;
    
    for (const auto& junction : junctions) {
        if (junction.circuit) {
            circuits[junction.circuit]++;
        }
    }
    
    std::vector<int> sizes;
    for (const auto& [circuit, size] : circuits) {
        sizes.push_back(size);
    }
    
    std::sort(sizes.begin(), sizes.end(), std::greater<int>());
    return sizes;
}

bool allConnected(const std::vector<Junction>& junctions) {
    auto circuit = junctions[0].circuit;
    if (!circuit) return false;
    
    for (const auto& junction : junctions) {
        if (junction.circuit != circuit) return false;
    }
    return true;
}

std::pair<std::vector<Junction>, std::vector<Pair>> parse(const std::string& input) {
    std::vector<Junction> junctions;
    std::istringstream iss(input);
    std::string line;
    
    while (std::getline(iss, line)) {
        if (line.empty()) continue;
        std::istringstream lineStream(line);
        std::string token;
        Junction j;
        
        std::getline(lineStream, token, ',');
        j.x = std::stoi(token);
        std::getline(lineStream, token, ',');
        j.y = std::stoi(token);
        std::getline(lineStream, token, ',');
        j.z = std::stoi(token);
        j.circuit = nullptr;
        
        junctions.push_back(j);
    }
    
    std::vector<Pair> pairs;
    for (size_t i = 0; i < junctions.size(); i++) {
        for (size_t j = i + 1; j < junctions.size(); j++) {
            Junction* a = &junctions[i];
            Junction* b = &junctions[j];
            double dist = std::sqrt(
                std::pow(a->x - b->x, 2) +
                std::pow(a->y - b->y, 2) +
                std::pow(a->z - b->z, 2)
            );
            pairs.push_back({a, b, dist});
        }
    }
    
    std::sort(pairs.begin(), pairs.end(), 
              [](const Pair& p1, const Pair& p2) { return p1.dist < p2.dist; });
    
    return {junctions, pairs};
}

void part1(const std::string& input) {
    int times = 1000;
    auto [junctions, pairs] = parse(input);
    
    for (int i = 0; i < times; i++) {
        connectPair(pairs[i], junctions);
    }
    
    auto sizes = circuitSizes(junctions);
    long long result = (long long)sizes[0] * sizes[1] * sizes[2];
    std::cout << result << std::endl;
}

void part2(const std::string& input) {
    auto [junctions, pairs] = parse(input);
    Pair* lastPair = nullptr;
    size_t idx = 0;
    
    while (!allConnected(junctions)) {
        lastPair = &pairs[idx];
        connectPair(pairs[idx], junctions);
        idx++;
    }
    
    long long ans = (long long)lastPair->a->x * lastPair->b->x;
    std::cout << ans << std::endl;
}

int main() {
    std::ifstream file("input.txt");
    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string input = buffer.str();
    
    part1(input);
    part2(input);
    
    return 0;
}